admin/server/device_code.go

```diff
@@ -0,0 +1,315 @@
+		return
+	}
+	verificationURI := s.conf.DeviceVerificationHost + "/oauth/device"
+	resp := DeviceCodeResponse{
+		DeviceCode:              authCode.DeviceCode,
+		UserCode:                authCode.UserCode,
+		VerificationURI:         verificationURI,
+		VerificationCompleteURI: verificationURI + "?user_code=" + authCode.UserCode,
+		ExpiresIn:               tokenExpirationSeconds,
+		PollingInterval:         5,
+	}
+	e := json.NewEncoder(w)
+	e.SetIndent("", "  ")
+	err = e.Encode(resp)
+	if err != nil {
+		s.logger.Error("failed to encode response", zap.Error(err))
+		internalServerError(w, err)
+	}
+}
+
+// handleUserCodeConfirmation handles the user code confirmation page. The user code is displayed
+// to the user and they are asked to confirm that they want to authorize the device. If the user
+// confirms, the device code is marked as approved in the server's device code store.
+func (s *Server) handleUserCodeConfirmation(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	userCode := r.URL.Query().Get("user_code")
+	if userCode == "" {
+		http.Error(w, "user_code is required", http.StatusBadRequest)
+		return
+	}
+	confirmation := r.URL.Query().Get("code_confirmed")
+	if confirmation == "" {
+		http.Error(w, "no code confirmation", http.StatusBadRequest)
+		return
+	}
+	sess, err := sessions.NewCookieStore([]byte(s.conf.SessionSecret)).Get(r, authSessionName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get session: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	var profiles map[string]interface{}
+	profile := sess.Values["profile"].([]byte)
+	err = json.Unmarshal(profile, &profiles)
+	if err != nil {
+		s.logger.Error("failed to unmarshal profile", zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	userID := profiles["sub"].(string)
+	if userID == "" {
+		s.logger.Error("no subject in profile: " + string(profile))
+		internalServerError(w, err)
+		return
+	}
+	authCode, err := s.admin.DB.GetByUserCode(r.Context(), userCode)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			http.Error(w, fmt.Sprintf("no such user code: %s found", userCode), http.StatusBadRequest)
+			return
+		}
+		s.logger.Error("failed to get auth code for userCode: "+userCode, zap.Error(err))
+		internalServerError(w, err)
+		return
+	}
+	if authCode.Used {
+		http.Error(w, "device code already used", http.StatusBadRequest)
+		return
+	}
+	if authCode.Expiry.Before(time.Now()) {
+		http.Error(w, "device code expired", http.StatusBadRequest)
+		return
+	}
+
+	// Update user code with user id and approval
+	authCode.UserID = userID
+	if confirmation != "true" {
+		authCode.ApprovalState = database.Rejected
+	} else {
+		authCode.ApprovalState = database.Approved
+	}
+	err = s.admin.DB.UpdateAuthCodeByUserCode(r.Context(), userCode, authCode)
+	if err != nil {
+		s.logger.Error("failed to update auth code for userCode: "+userCode, zap.Error(err))
+		internalServerError(w, err)
+	}
+}
```

Comment:
Maybe move this logic into something like `s.admin.ConfirmAuthCode(ctx context.Context, userCode string, confir, bool) error`

Reply:
you mean entire logic of `handleUserCodeConfirmation` ?

Reply:
I was just thinking the highlighted part here – basically, `admin.Service` is for business logic, and `admin/server` is for networking-related things (status codes, cookies, serializing/deserializing requests/responses).